sycl/include/CL/sycl/queue.hpp

```diff
@@ -70,56 +123,111 @@ class queue {
+  /// queue.
+  ///
+  /// Synchronous errors will be reported through SYCL exceptions. Asynchronous
+  /// errors will be passed to the async_handler passed to the queue on
+  /// construction. If no async_handler was provided then asynchronous
+  /// exceptions will be lost.
+  void wait_and_throw();
+
+  /// Checks if any asynchronous errors have been produced by the queue and if
+  /// so reports them to the async_handler passed on the queue construction.
+  ///
+  /// If no async_handler was provided then asynchronous exceptions will be
+  /// lost.
+  void throw_asynchronous();
 
-  event memcpy(void* Dest, const void* Src, size_t Count) {
-    return impl->memcpy(impl, Dest, Src, Count);
-  }
+  /// @return true if the queue was constructed with property specified by
+  /// PropertyT.
+  template <typename PropertyT> bool has_property() const;
 
-  event mem_advise(const void *Ptr, size_t Length, int Advice) {
-    return impl->mem_advise(Ptr, Length, Advice);
-  }
+  /// @return a copy of the property of type PropertyT that the queue was
+  /// constructed with. If the queue was not constructed with the PropertyT
+  /// property, an invalid_object_error SYCL exception.
+  template <typename propertyT> propertyT get_property() const;
 
+  /// Fills the memory pointed by a USM pointer with the value specified.
+  ///
+  /// @param Ptr is a USM pointer to the memory to fill.
+  /// @param Value is a value to be set. Value is cast as an unsigned char.
+  /// @param Count is a number of bytes to fill.
+  /// @return an event representing fill operation.
+  event memset(void *Ptr, int Value, size_t Count);
+
+  /// Copies data from one memory region to another, both pointed by
+  /// USM pointers.
+  ///
+  /// @param Dest is a USM pointer to the destination memory.
+  /// @param Src is a USM pointer to the source memory.
+  /// @param Count is a number of bytes to copy.
+  /// @return an event representing copy operation.
+  event memcpy(void *Dest, const void *Src, size_t Count);
+
+  /// Provides additional information to the underlying runtime about how
+  /// different allocations are used.
+  ///
+  /// @param Ptr is a USM pointer to the allocation.
+  /// @param Length is a number of bytes in the allocation.
+  /// @param Advice is a device-defined advice for the specified allocation.
+  /// @return an event representing advice operation.
+  event mem_advise(const void *Ptr, size_t Length, int Advice);
+
+  /// Provides hints to the runtime library that data should be made available
+  /// on a device earlier than Unified Shared Memory would normally require it
+  /// to be available.
+  ///
+  /// @param Ptr is a USM pointer to the memory to be prefetched to the device.
+  /// @param Count is a number of bytes to be prefetched.
   event prefetch(const void* Ptr, size_t Count) {
-    return submit([=](handler &CGH) {
-        CGH.prefetch(Ptr, Count);
-    });
+    return submit([=](handler &cgh) { cgh.prefetch(Ptr, Count); });
```

Comment:
Why don't move it to .cpp ?

Reply:
@romanovvlad It's a harmless one-liner, let compiler just inline it.